File: javascript/src/problems101_200/word-ladder.js
class Problem {
  ladderLength(beginWord, endWord, wordList) {
    throw new Error(`Not implemented (${beginWord}, ${endWord}, ${wordList}).`);
  }
}

class Default extends Problem {
  ladderLength(beginWord, endWord, wordList) {
    if (!wordList.includes(endWord)) {
      return 0;
    }

    const wordSet = new Set(wordList);

    const queue = [];
    queue.push(beginWord);
    let steps = 0;

    while (queue.length > 0) {
      const size = queue.length;
      for (let i = 0; i < size; i++) {
        const s = queue.shift();
        if (s === endWord) {
          return steps + 1;
        }
        for (let j = 0; j < s.length; j++) {
          const word = s.split('');
          for (let c = 'a'.charCodeAt(0); c <= 'z'.charCodeAt(0); c++) {
            word[j] = String.fromCharCode(c);
            const newWord = word.join('');
            if (s === newWord || !wordSet.has(newWord)) {
              continue;
            }
            queue.push(newWord);
            wordSet.delete(newWord);
          }
        }
      }
      steps++;
    }
    return 0;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems101_200/word_ladder.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def ladder_length(self, begin_word: str, end_word: str, word_list: list[str]) -> int:
        pass


class Default(Problem):
    @override
    def ladder_length(self, begin_word: str, end_word: str, word_list: list[str]) -> int:
        if end_word not in word_list:
            return 0

        # Python's set has O(1) lookup time
        word_set: set[str] = set(word_list)

        queue: list[str] = [begin_word]
        steps: int = 0
        while queue:
            for _ in range(len(queue)):
                s: str = queue.pop(0)
                if s == end_word:
                    return steps + 1
                for j in range(len(s)):
                    word: list[str] = list(s)
                    for c in range(ord('a'), ord('z') + 1):
                        word[j] = chr(c)
                        new_word: str = ''.join(word)
                        if s == new_word or new_word not in word_set:
                            continue
                        queue.append(new_word)
                        word_set.remove(new_word)
            steps += 1

        return 0


SOLUTIONS: list[Problem] = [Default()]
